packages/plugins/src/mf.ts

```diff
@@ -0,0 +1,247 @@
+    config: {
+      schema(Joi) {
+        return Joi.object({
+          name: Joi.string(),
+          fieldName: Joi.string().default('name'),
+          remotes: Joi.array().items(
+            Joi.object({
+              aliasName: Joi.string(),
+              name: Joi.string().required(),
+              entry: Joi.string(),
+              entries: Joi.object(),
+              keyResolver: Joi.string(),
+            }),
+          ),
+          library: Joi.object(),
+        });
+      },
+    },
+    enableBy: api.EnableBy.config,
+  });
+
+  api.modifyWebpackConfig(async (config, { webpack }) => {
+    const exposes = await constructExposes();
+    const remotes = formatRemotes();
+    const shared = mergeSharedConfig();
+
+    // @ts-ignore
+    if (isEmpty(remotes) && isEmpty(exposes)) {
+      api.logger.warn(
+        'ModuleFederation exposes and remotes are empty, plugin will not work',
+      );
+      return config;
+    }
+
+    if (!isEmpty(remotes)) {
+      if (!(api.env === 'development' && api.config.mfsu)) {
+        changeUmiEntry(config);
+      }
+    }
+
+    let name = '_';
+
+    if (!isEmpty(exposes)) {
+      name = mfName();
+
+      addMFEntry(
+        config,
+        name,
+        join(api.paths.absTmpPath, 'plugin-mf', mfSetupModuleVirtualPath),
+      );
+    }
+
+    const mfConfig = {
+      name,
+      remotes,
+      filename: 'remote.js',
+      exposes,
+      shared,
+      library: api.config.mf.library,
+    };
+
+    const { ModuleFederationPlugin } = webpack.container;
+    // @ts-ignore
+    config.plugins.push(new ModuleFederationPlugin(mfConfig));
+    api.logger.debug(
+      `ModuleFederationPlugin is enabled with config ${JSON.stringify(
+        mfConfig,
+      )}`,
+    );
+
+    return config;
+  });
+
+  api.onGenerateFiles(() => {
+    api.writeTmpFile({
+      content: `/* infer remote public */;
+      __webpack_public_path__ = document.currentScript.src + '/../';`,
+      path: mfSetupModuleVirtualPath,
+    });
+
+    if (api.env === 'development' && api.config.mfsu) {
+      // skip mfsu already dynamic import
+      return;
+    }
+
+    api.writeTmpFile({
+      noPluginDir: true,
+      content: `import('${winPath(join(api.paths.absTmpPath, 'umi.ts'))}')`,
+      path: 'asyncEntry.ts',
+    });
+  });
+
+  function formatRemotes() {
+    const { remotes = [] } = api.config.mf;
+
+    const memo: Record<string, string> = {};
+
+    remotes.forEach((remote: any) => {
+      const aliasName = remote.aliasName || remote.name;
+
+      const r = formatRemote(remote);
+
+      if (memo[aliasName]) {
+        return api.logger.error(
+          `${aliasName} already set as ${memo[aliasName]}, new value ${r} will be ignored`,
+        );
+      }
+      memo[aliasName] = r;
+    });
+
+    return memo;
+  }
+
+  function formatRemote(remote: any): string {
+    const url = remote.entry;
+
+    const mfUrl = `${remote.name}@${url}`;
+
+    if (remote.entry) {
+      return mfUrl;
```

Comment:
```suggestion
      return `${remote.name}@${remote.entry}`
```

上面 `url` 、`mfUrl` 这两个变量都没再用到，去掉他们，这样写会不更好。